### ISSUE TITLE
Support more than 64K pixels

### ISSUE THREAD

**Issue by marcmerlin** [FastLED/FastLED#968]
One thing that's been biting me recently is as you may remember, I'm offering a FastLED CRGB backed framebuffer that displays on lots of backends, from addressable pixels, to RGBPanels, to TFT screens ( https://github.com/marcmerlin/Framebuffer_GFX ).

One issue I'm hitting with TFT screens and RGBPanels on rPi (I wrote a glue library to run Arduino and FastLED sketches on rPi  https://github.com/marcmerlin/FastLED_RPIRGBPanel_GFX ).
Long story short, I'm hitting the 65536 pixel limit that is assumed in multiple FastLED functions like fade and friends.

Would it be possible to raise all those uint16_t to uint32_t ?

**Comment by kriegsman** [comment]
You know, I've thought about this, and I think perhaps the thing to do is to create a typedef something like pixelIndex_t which could default to 16 bits wide on 8-bit (AVR) platforms, default to 32 bits on 32-bit (ARM, ESP, Intel, etc) platforms, and in all cases be user-overridable with a #define.

**Comment by marcmerlin** [comment]
Thanks @kriegsman 
I just built a new RGBPanel array that's 384*192, turns out that's also too many pixels :)
I realize you're trying to save a few precious bytes for people still using 328p. If you are worried about the few extra bytes used by the functions that will process 32bit instead of 16bit?  I think the caller can still send and receive 16bit so that it remains compatible?
I can however see that if a function returns a 32bit int to a caller that stores the result in a 16bit int, that would trigger a compilation warning, but still work as long as there is no overflow, correct?
That said, if you are ok with the extra work of making the function change between 16 and 32bit depending on the the CPU, that'd be fine with me, or use "int" which should default to 16bit on 328p and 32bit on better MCUs?

**Comment by bbulkow** [comment]
In the days when arduinos had 2k of RAM, this made sense, but those days are vanishing. With our greatly little embedded systems cresting into megabytes, and this code being used on higher power machines like the RPI, I think it's time to raise the type. Alternately, since someone's going in, use an abstract type ( PixelLen_t ) and thus allow different ports to change it.

I'm not above 64k yet because each of the ESP32 controllers I use is only 10 bucks and well networked, but I can see larger panels where it makes sense.

**Comment by zackees** [comment]
I will accept a pr to change this.

**Comment by ewoudwijma** [comment]
> I will accept a pr to change this.

I am away from keyboard so can't oversee the memory foodprint consequences of this change but can you check impact as if this changes stack or heap too much in heavy use cases of FastLED it would be a pity if we run out of memory

**Comment by zackees** [comment]
The memory footprint will be negligible, we are just widening the index variable located on the stack by 2 bytes. The issue is more of the speed on the avr boards because of tight timing requirements for bit banging out the WS2812 leds.

**Comment by zackees** [comment]
Deferred by automated triage: Widening the pixel-index type is a public-API change across ~25 functions, and the two maintainer proposals on this thread (kriegsman's conditional pixelIndex_t vs zackees' flat widening) disagree on what AVR should get. Needs a maintainer decision before implementation. Full survey of the affected surface posted on the issue.

**Comment by zackees** [comment]
Surveyed the current tree to scope this. Marking `deferred` — not because it is hard, but because the one decision it hinges on has two conflicting maintainer answers on this very thread. Posting the survey so whoever picks it up does not have to redo it.

## The good news: the output path is already 32-bit clean

The 64K ceiling is **not** in the controller/driver stack. Every length there is already `int`:

| Location | Type |
|---|---|
| `FastLED.h:708` `addLeds(..., int nLedsOrOffset, int nLedsIfOffset)` | `int` |
| `fl/channels/cled_controller.h:100,106,270` `showColor`/`show`/`setLeds` | `int nLeds` |
| `fl/channels/cled_controller.h:298` `CLEDController::size()` | `int` |
| `pixel_controller.h:104` `PixelController::mLen` | `int` |
| `pixel_controller.h:383` `PixelController::size()` | `int` |
| `pixelset.h:105` `CPixelView::len` | `const int` |

So on any 32-bit target you can already *drive* a >65535-pixel buffer today. @marcmerlin's 384×192 = 73,728 panel is fine as far as `show()` is concerned.

## The actual ceiling: the color-utility layer

Exactly as reported — "fade and friends". Raw-pointer overloads still take `fl::u16`:

- **`fl/gfx/colorutils.h`** — `fadeLightBy` (57), `fade_video` (65), `nscale8_video` (77), `fadeToBlackBy` (89), `fade_raw` (97), `nscale8` (110), `fadeUsingColor` (128), `blend` (179, 191), `nblend` (213, 225)
- **`fl/gfx/fill.h`** — `fill_gradient` (105, 216, 232, 251), `fill_gradient_RGB` (274, 283, 293, 304)
- **`fl/gfx/blur.h`** — `blur1d` (33, 178)
- **`power_mgt.h`** — `calculate_unscaled_power_mW` (256), `calculate_max_brightness_for_power_mW` (276), `calculate_max_brightness_for_power_vmA` (286)

## The part that is actively worse than a plain limit

The newer `fl::span` convenience overloads **silently truncate**. There are **14 explicit `static_cast<fl::u16>(...)` in `colorutils.h` and 3 in `fill.h`** of exactly this shape:

```cpp
inline void fadeToBlackBy(fl::span<CRGB> leds, fl::u8 fadeBy) FL_NO_EXCEPT {
    fadeToBlackBy(leds.data(), static_cast<fl::u16>(leds.size()), fadeBy);
}
```

Hand that a 73,728-pixel span and it fades the first **8,192** and returns cleanly. The explicit cast means no compiler warning fires. That is a wrong-answer bug, not a documented limit, and it is arguably worth fixing on its own even if the wider retype stalls.

Note the newest APIs already got this right — `calculate_unscaled_power_mW(fl::span<const CRGB>)` at `power_mgt.cpp.hpp:158` iterates with `size_t`.

## ⚠️ Trap for anyone reaching for sed

`fl::u16` in this file is heavily overloaded and most instances **must not** be widened:
- `ColorFromPaletteExtended(pal, fl::u16 index, ...)`, `ColorFromPaletteHD` — that `u16` is a **fixed-point fraction** across the palette, not a count.
- `CHSVPalette::fillSolid(CHSV*, fl::u16 size, ...)`, `template <typename TColor, fl::u16 Size>` — **palette** sizes (16/32/256), not pixel counts.

A blind `u16`→`u32` sweep over `colorutils.h` silently changes palette interpolation math. The counts and the fractions have to be separated by hand.

## The blocking decision

@kriegsman (2020): a `pixelIndex_t` typedef, 16-bit on AVR, 32-bit elsewhere, user-overridable by `#define`.
@zackees (2024): "The memory footprint will be negligible... The issue is more of the speed on the avr boards."

These give **different answers for AVR**, and AVR is the only target where it matters — on a 328p, 2KB of SRAM caps you at ~682 pixels, so 32-bit indices there are pure cost (cycles and flash) for a range that is physically unreachable. That argues for @kriegsman's conditional typedef. But the conditional makes the public signatures typedef-dependent, which is a real API-surface commitment.

**Someone with commit rights needs to pick one.** Once picked, the change is mechanical and I am happy to do it. Concretely:

1. Which? (a) conditional `fl::led_index_t` — 16 on AVR, 32 elsewhere, `#define`-overridable, or (b) flat `fl::u32` everywhere.
2. Does `CPixelView::len` / `CLEDController` move off `int` too? On AVR `int` is 16-bit, so `CRGBSet` is already capped at 32,767 there regardless of what happens to the utility functions.
3. Ship the span-truncation fix separately and first? It is a strictly-correct bugfix with no design content, and it unblocks the reported use case for anyone already passing spans.

Happy to take direction on any of the three.
